File: python/sparklabs/batch/sales/dataset/product_rank.py
from pyspark.sql import Window
from pyspark.sql import functions as F

from sparklabs.batch.sales.dataset import utils


def run() -> None:
    window_spec = Window.orderBy(F.desc("category_count"))

    (
        utils.order_items.groupBy(F.col("product_id"))
        .agg(F.count("*").alias("order_count"))
        .join(utils.products, ["product_id"], "inner")
        .groupBy(F.col("category"))
        .agg(F.count("*").alias("category_count"))
        .withColumn("category_rank", F.dense_rank().over(window_spec))
        .show(truncate=False)
    )


if __name__ == "__main__":
    run()

File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/dataset/ProductRank.scala
package io.devorbit.sparklabs.batch.sales.dataset

import org.apache.spark.sql.expressions.Window
import org.apache.spark.sql.functions.{col, count, dense_rank, desc}

object ProductRank extends App {
  val windowSpec = Window.orderBy(desc("category_count"))

  DatasetUtils.toDataFrame(DatasetUtils.orderItems)
    .groupBy(col("product_id"))
    .agg(count("*").as("order_count"))
    .join(DatasetUtils.toDataFrame(DatasetUtils.products), Seq("product_id"), "inner")
    .groupBy(col("category"))
    .agg(count("*").as("category_count"))
    .withColumn("category_rank", dense_rank().over(windowSpec))
    .show(false)
}
